python_scripts/linear_models_feature_engineering_classification.py

```diff
@@ -0,0 +1,437 @@
+# ---
+# jupyter:
+#   kernelspec:
+#     display_name: Python 3
+#     name: python3
+# ---
+
+# %% [markdown]
+#
+# # Non-linear feature engineering for Logistic Regression
+#
+# Linear classification models can also be used for classification on
+# non-linearly separable datasets. To do so, we transform the original feature
+# feature space into a new feature space using a non-linear transformation. The
+# resulting feature space has typically more dimensions than the original
+# feature space, and as a result, the linear model can separate the data more
+# easily.
+#
+# Let us illustrate this on three synthetic datasets. Each dataset has two
+# original features and two classes to make it easy to visualize. The first
+# dataset is called the "moons" dataset as the data points from each class are
+# shaped as a crescent moon:
+
+# %%
+import numpy as np
+import pandas as pd
+from sklearn.datasets import make_moons
+
+feature_names = ["Feature #0", "Feature #1"]
+target_name = "class"
+
+X, y = make_moons(n_samples=100, noise=0.13, random_state=42)
+
+# We store both the data and target in a dataframe to ease plotting
+moons = pd.DataFrame(
+    np.concatenate([X, y[:, np.newaxis]], axis=1),
+    columns=feature_names + [target_name],
+)
+data_moons, target_moons = moons[feature_names], moons[target_name]
+
+# %% [markdown]
+#
+# The second dataset is called the "Gaussian quantiles" dataset as the data
+# points are sampled from the same (Gaussian) distribution in a 2D space with a
+# higher density of points in the center. The points closest to the center are
+# assigned to the class 1 while the points in the outer edges are assigned to
+# the class 0.
+
+# %%
+from sklearn.datasets import make_gaussian_quantiles
+
+X, y = make_gaussian_quantiles(
+    n_samples=100, n_features=2, n_classes=2, random_state=42
+)
+gauss = pd.DataFrame(
+    np.concatenate([X, y[:, np.newaxis]], axis=1),
+    columns=feature_names + [target_name],
+)
+data_gauss, target_gauss = gauss[feature_names], gauss[target_name]
+
+# %% [markdown]
+#
+# The third dataset is called the "XOR" dataset as the data points are sampled
+# from a uniform distribution in a 2D space and the class is defined by the
+# Exclusive OR (XOR) operation on the two features: the target class is 1 if
+# only one of the two features is greater than 0. The target class is 0
+# otherwise.
+
+# %%
+xor = pd.DataFrame(
+    np.random.RandomState(0).uniform(low=-1, high=1, size=(200, 2)),
+    columns=feature_names,
+)
+target_xor = np.logical_xor(xor["Feature #0"] > 0, xor["Feature #1"] > 0)
+target_xor = target_xor.astype(np.int32)
+xor["class"] = target_xor
+data_xor = xor[feature_names]
+
+# %% [markdown]
+#
+# We use matplotlib to visualized the data points in both datasets:
+
+# %%
+import matplotlib.pyplot as plt
+import seaborn as sns
+import warnings
+
+
+warnings.filterwarnings("ignore", module="seaborn")
```

Comment:
I noticed seaborn is raising FutureWarning in several places inside the notebook. I am wondering if the best solution is to ping to a previous version. But at least we should mention why we are adding this line, WDYT?

Reply:
I agree, those warnings are annoying. But rather than pinning seaborn, we might have to pin an older version of pandas. Or both.

Reply:
Or maybe we can put this call to `warnings.filterwarnings("ignore", module="seaborn")` outside of the notebooks themselves, maybe in `conf.py` or something similar if sphinx or jupyter book allow for this kind of thing.

Reply:
I removed the warnings filter in 1338b95.